services/gitdiff/gitdiff.go

```diff
@@ -1094,13 +1092,16 @@ func readFileName(rd *strings.Reader) (string, bool) {
 // DiffOptions represents the options for a DiffRange
 type DiffOptions struct {
 	BeforeCommitID     string
+	BeforeCommit       *git.Commit
 	AfterCommitID      string
+	AfterCommit        *git.Commit
```

Comment:
How do these 2 new fields help the performance?

It seems that there is no loop to use the commit IDs in `DiffOptions`, then does filling `BeforeCommit` ahead really improve performance?

Reply:
The caller will need to prefill these to avoid the load. Before this change, this was in `diffFile.GetTailSection`, which does run in a loop, but behind a conditional.

Reply:
In some cases, the caller of `GetDiff` already has loaded this, so we'd want to avoid looking it up. This happens in when viewing a pull reques. See https://github.com/go-gitea/gitea/pull/32393/files#diff-9107a2d87c7913bc4fbdf735c0391d8e9ed5936ebac72a9bd23188616accc4a7R759

Reply:
This `GetDiff` case, it is not in a loop and it should be quite fast?

I would prefer to keep the logic simple, avoid bloating the `DiffOptions` struct .....

Reply:
fixed in https://github.com/go-gitea/gitea/pull/32393/commits/3faa3c68459fbe9bb99332acdf25d3d2fc898939